onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon_int8.cpp

```diff
@@ -0,0 +1,1315 @@
+    size_t StrideQuantBScale,
+    size_t StrideQuantBZeroPoint,
+    const std::byte*& QuantBDataColPtr,
+    const float*& QuantBScaleColPtr,
+    const std::byte*& QuantBZeroPointColPtr,
+    const float*& BiasPtr,
+    float*& SumPtr
+)
+{
+    QuantBDataColPtr += NumCols * StrideQuantBData;
+    QuantBScaleColPtr += NumCols * StrideQuantBScale;
+    if constexpr (HasZeroPoint) {
+        QuantBZeroPointColPtr += NumCols * StrideQuantBZeroPoint;
+    }
+
+    BiasPtr += BiasPtr != nullptr ? NumCols : 0;
+    SumPtr += NumCols;
+}
+
+template <size_t NumRows>
+MLAS_FORCEINLINE void
+AdvanceRowPtrs(
+    size_t StrideQuantA,
+    size_t ldc,
+    const std::byte*& QuantARowPtr,
+    float*& SumRowPtr
+)
+{
+    QuantARowPtr += NumRows * StrideQuantA;
+    SumRowPtr += NumRows * ldc;
+}
+
+template <bool HasZeroPoint>
+void
+SQ4BitGemmKernel_CompInt8_BlkLen16(
+    const std::byte* QuantA,
+    const std::byte* QuantBData,
+    const float* QuantBScale,
+    const std::byte* QuantBZeroPoint,
+    float* C,
+    size_t CountM,
+    size_t CountN,
+    size_t BlockCountK,
+    size_t ldc,
+    const float* Bias
+)
+{
+    constexpr size_t BlkBitWidth = 4;
+    constexpr size_t BlkLen = 16;
+
+    const size_t StrideQuantA = BlockCountK * Q8BlkSize(BlkLen);
+
+    const size_t StrideQuantBData = BlockCountK * MlasQNBitBlkDataSizeInBytes(BlkBitWidth, BlkLen);
+    const size_t StrideQuantBScale = BlockCountK;
+    const size_t StrideQuantBZeroPoint = MlasQNBitZeroPointsForBlksSizeInBytes<BlkBitWidth>(BlockCountK);
+
+    const std::byte* QuantARowPtr = QuantA;
+
+    float* SumRowPtr = C;
+
+    size_t m_remaining = CountM;
+    while (m_remaining > 1) {
+        const std::byte* QuantBDataColPtr = QuantBData;
+        const float* QuantBScaleColPtr = QuantBScale;
+        const std::byte* QuantBZeroPointColPtr = QuantBZeroPoint;
+
+        const float* BiasPtr = Bias;
+
+        float* SumPtr = SumRowPtr;
+
+        size_t n_remaining = CountN;
+        while (n_remaining > 1) {
+            // Compute 2x2 tiles of output
+            SQ4BitGemm_CompInt8_Compute2x2_BlkLen16<HasZeroPoint>(
+                QuantARowPtr,
+                QuantBDataColPtr,
+                QuantBScaleColPtr,
+                QuantBZeroPointColPtr,
+                BiasPtr,
+                SumPtr,
+                BlockCountK,
+                StrideQuantA,
+                StrideQuantBData,
+                StrideQuantBScale,
+                StrideQuantBZeroPoint,
+                ldc
+            );
+
+            // Move to next 2 columns
+            AdvanceColPtrs<2, HasZeroPoint>(
+                StrideQuantBData, StrideQuantBScale, StrideQuantBZeroPoint,
+                QuantBDataColPtr, QuantBScaleColPtr, QuantBZeroPointColPtr, BiasPtr, SumPtr
+            );
+
+            n_remaining -= 2;
+        }
+
+        if (n_remaining > 0) {
+            // Compute last 2x1 tile of output
+            SQ4BitGemm_CompInt8_Compute1x1_BlkLen16<HasZeroPoint>(
```

Comment:
nip. Is there a need for 2x1? Data from matrix B can be loaded and shared with 2 rows of A. I understand this will not have much of impact. Maybe avx shall follow this to reduce code size.

Reply:
It might help. However, it's limited to where N is odd, and then it only would run for the last column. I'm inclined to keep it simpler for now.

We can add other tile sizes later on if there's a significant benefit. I think 2x4 or 2x3 would also be worth exploring.